tox.ini

```diff
@@ -42,11 +42,22 @@ commands =
 
 [testenv:integrationtests]
 basepython = python3.8
+passenv = LANG TERM LANGUAGE LC_ALL LD_LIBRARY_PATH SOLC_BINARY BRANCH INTEGRATION_TESTS_REPO_URL
 deps =
 	-rrequirements/tests.txt
+setenv =
+	INT_TESTS_REPO_SSH_URL = git@github.com:gridsingularity/gsy-backend-integration-tests.git
+whitelist_externals =
+	git
+	bash
+commands_pre =
+	pip install -e .
+	git clone {env:INTEGRATION_TESTS_REPO_URL:{env:INT_TESTS_REPO_SSH_URL:}} {envtmpdir}/gsy-backend-integration-tests
 commands =
-    pip install -e .
-	behave --tags=-disabled integration_tests {posargs}
+	bash -c 'cp -r {envtmpdir}/gsy-backend-integration-tests/gsy_e_tests/integration_tests/ ./integration_tests/'
+	bash -c 'for f in ./integration_tests/*.feature; do behave --tags=-slow --tags=-disabled "$f" || exit 1; done'
```

Comment:
What's the difference between this and the following?
```
bash -c 'behave --tags=-slow --tags=-disabled integration_tests'
```

Reply:
This implementation executes each `*.feature` file one by one, so if any integration test fail on the` *.feature` file being executed, the bash command exits and therefore testenv (ci/integrationtests) fails. If the tests on `*.feature` file pass, then it executes the following `*.feature` file on the directory.

On the command you posted, the tests are executed one after the other, regardless of wether they failed or not.

I guess the first approach saves resources, interrupting the process after the first test fails. The second one executes all of them. Both testenv approaches would fail at the end. 
I guess one of them is just somehow more efficient.

Reply:
A downside for the first approach is that you could not get the information of all of the tests that failed due to changes. You would just know something broke, maybe fix it and then run it again until you find out if anything different fails, and so on.

I took the first approach because it was already implemented when I started working on this ticket, but I'm ok with both of them. Actually I ocasionaly used the second just to save me some time on identifying all of the failing tests...

Reply:
In my personal preference and workflow, I would actually go for the second one. As you said, I would find it better to see all failures at once to be able to fix them. Moreover, it might be that one single test doesn't give me enough information to understand what's going wrong, while seeing all the failing tests might provide more context and information to infer the real cause of the failures. @hannesdiedrich what do you think of this?

Reply:
The early exit was implemented by spyros [here](https://github.com/gridsingularity/gsy-e/commit/91339e147f3074c3a33a32b6159ca230b3ebd6a4) and the explanation is [here](https://github.com/gridsingularity/gsy-e/pull/1363#issue-1110512892).
Basically the runtime is much lower if you early return.
I agree with that. If you want to debug the failing intehgration tests, you can run `behave --tags=-slow --tags=-disabled integration_tests` just locally IMO.

Reply:
I'm not totally convinced, but it's just because I have a different personal preference. So it's fine to keep it as it is. Just one thing: behave already provides this feature out of the box ([`--stop`](https://behave.readthedocs.io/en/stable/behave.html#cmdoption-stop)), so I think the code could be simplified to just be:
```bash
behave --tags=-slow --tags=-disabled integration_tests --stop
```
I haven't tried it but it's supposed to work.

Reply:
I agree with all of you, that from the tester's perspective he would like to see all integration tests that actually fail after a CI test run. However, from the CI worker's perspective, exiting on the first integration test that fails will save a lot of the computation time. Keep in mind we are talking only about the gsy-e integration tests, which take around 15 mins on a CI worker, and if the worker exits as soon as one of the first integration tests fail, it will leave the worker free for other test runs that are blocked during this time. 

The reason for creating this bash command, was essentially memory leaks, so my focus was not on the behave env but more on the Python interpreter that was running it. After inspecting the memory usage of behave locally, I saw that it was increasing gradually while the integration tests were running. This had the effect that the first integration tests were relatively fast, but the following tests were gradually becoming slower. After some time inspecting the behave internals (during which I did not find any variables that were not deallocated from us) I decided to follow the easy solution and recreate the behave / Python interpreter at the end of every feature file, in order to deallocate the memory and start from scratch. 

Keep in mind that these 2 issues that are discussed are not coupled, meaning that even the bash script can be adapted to run all the tests despite some feature files failing.

Reply:
Thanks for the explanation Spyros! Regarding the first point, I would then suggest using the "early stop" command only for the `ci` environment and removing it from the other environments that are not used by the CI. 
(Minor point, related to this: maybe our tox file could be refactored a bit. I have the feeling that we could reuse many parts of it to avoid code repetition. Not too important anyway).